sql/core/src/main/scala/org/apache/spark/sql/execution/basicPhysicalOperators.scala

```diff
@@ -228,6 +234,156 @@ case class FilterExec(condition: Expression, child: SparkPlan)
+ *
+ * Anchor terms are physical plans and they are used to initialize the query in the first run.
+ * Recursive terms are used to extend the result with new rows, They are logical plans and contain
+ * references to the result of the previous iteration or to the so far cumulated result. These
+ * references are updated with new statistics and compiled to physical plan and then updated to
+ * reflect the appropriate RDD before execution.
+ *
+ * The execution terminates once the anchor terms or the current iteration of the recursive terms
+ * return no rows or the number of cumulated rows reaches the limit.
+ *
+ * During the execution of a recursive query the previously computed results are reused multiple
+ * times. To avoid massive recomputation of these pieces of the final result, they are cached.
+ *
+ * @param name the name of the recursive table
+ * @param anchorTerms this child is used for initializing the query
+ * @param recursiveTerms this child is used for extending the set of results with new rows based on
+ *                       the results of the previous iteration (or the anchor in the first
+ *                       iteration)
+ * @param limit the maximum number of rows to return
+ */
+case class RecursiveTableExec(
+    name: String,
+    anchorTerms: Seq[SparkPlan],
+    @transient
+    val recursiveTerms: Seq[LogicalPlan],
+    limit: Option[Long]) extends SparkPlan {
+  override def children: Seq[SparkPlan] = anchorTerms
+
+  override def output: Seq[Attribute] = anchorTerms.head.output.map(_.withNullability(true))
+
+  override def simpleString(maxFields: Int): String =
+    s"RecursiveTable $name${limit.map(", " + _).getOrElse("")}"
+
+  override def innerChildren: Seq[QueryPlan[_]] = recursiveTerms ++ super.innerChildren
+
+  override protected def doExecute(): RDD[InternalRow] = {
+    val storageLevel = StorageLevel.fromString(conf.getConf(SQLConf.RECURSION_CACHE_STORAGE_LEVEL))
+
+    val prevIterationRDDs = ArrayBuffer.empty[RDD[InternalRow]]
+    var prevIterationCount = 0L
+
+    val anchorTermsIterator = anchorTerms.iterator
+    while (anchorTermsIterator.hasNext && limit.forall(_ > prevIterationCount)) {
+      val anchorTerm = anchorTermsIterator.next()
+
+      lazy val cumulatedResult = if (prevIterationRDDs.size > 1) {
+        sparkContext.union(prevIterationRDDs)
+      } else {
+        prevIterationRDDs.head
+      }
+
+      anchorTerm.foreach {
+        case rr: RecursiveReferenceExec if rr.name == name => rr.recursiveTable = cumulatedResult
+        case _ =>
+      }
+
+      val rdd = anchorTerm.execute().map(_.copy()).persist(storageLevel)
+      val count = rdd.count()
+      if (count > 0) {
+        prevIterationRDDs += rdd
+        prevIterationCount += count
+      }
+    }
+
+    val cumulatedRDDs = ArrayBuffer(prevIterationRDDs: _*)
+    var cumulatedCount = prevIterationCount
+    var level = 0
+    val levelLimit = conf.getConf(SQLConf.RECURSION_LEVEL_LIMIT)
```

Comment:
`conf.recursionLevelLimit`? Also, can you move this definition outside the loop?